flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java

```diff
@@ -77,63 +86,96 @@ public GenericWriteAheadSink(CheckpointCommitter committer, TypeSerializer<IN> s
+		Preconditions.checkNotNull(this.pendingHandles, "The operator has not been properly initialized.");
+
 		//only add handle if a new OperatorState was created since the last snapshot
 		if (out != null) {
 			StreamStateHandle handle = out.closeAndGetHandle();
-			if (state.pendingHandles.containsKey(checkpointId)) {
+
+			PendingCheckpointId pendingCheckpoint = new PendingCheckpointId(
+				checkpointId, getRuntimeContext().getIndexOfThisSubtask());
+
+			if (pendingHandles.containsKey(pendingCheckpoint)) {
 				//we already have a checkpoint stored for that ID that may have been partially written,
 				//so we discard this "alternate version" and use the stored checkpoint
 				handle.discardState();
 			} else {
-				state.pendingHandles.put(checkpointId, new Tuple2<>(timestamp, handle));
+				this.pendingHandles.put(pendingCheckpoint, new PendingHandle(timestamp, handle));
 			}
 			out = null;
 		}
 	}
 
 	@Override
-	public void snapshotState(FSDataOutputStream out,
-			long checkpointId,
-			long timestamp) throws Exception {
+	public void snapshotState(FSDataOutputStream out, long checkpointId, long timestamp) throws Exception {
 		saveHandleInState(checkpointId, timestamp);
 
-		InstantiationUtil.serializeObject(out, state);
+		DataOutputViewStreamWrapper outStream = new DataOutputViewStreamWrapper(out);
+		outStream.writeInt(pendingHandles.size());
+		for (Map.Entry<PendingCheckpointId, PendingHandle> pendingCheckpoint : pendingHandles.entrySet()) {
+			pendingCheckpoint.getKey().serialize(outStream);
+			pendingCheckpoint.getValue().serialize(outStream);
+		}
 	}
 
 	@Override
 	public void restoreState(FSDataInputStream in) throws Exception {
-		this.state = InstantiationUtil.deserializeObject(in, getUserCodeClassloader());
+		final DataInputViewStreamWrapper inStream = new DataInputViewStreamWrapper(in);
+		int noOfPendingHandles = inStream.readInt();
+		for (int i = 0; i < noOfPendingHandles; i++) {
+			PendingCheckpointId id = PendingCheckpointId.restore(inStream);
+			PendingHandle handle = PendingHandle.restore(inStream, getUserCodeClassloader());
+			this.pendingHandles.put(id, handle);
+		}
 	}
 
-	private void cleanState() throws Exception {
-		synchronized (this.state.pendingHandles) { //remove all handles that were already committed
-			Set<Long> pastCheckpointIds = this.state.pendingHandles.keySet();
-			Set<Long> checkpointsToRemove = new HashSet<>();
-			for (Long pastCheckpointId : pastCheckpointIds) {
-				if (committer.isCheckpointCommitted(pastCheckpointId)) {
-					checkpointsToRemove.add(pastCheckpointId);
+	/**
+	 * Called at {@link #open()} to clean-up the pending handle list.
+	 * It iterates over all restored pending handles, checks which ones are already
+	 * committed to the outside storage system and removes them from the list.
+	 */
+	private void cleanRestoredHandles() throws Exception {
+		synchronized (pendingHandles) {
+
+			Set<PendingCheckpointId> checkpointIdsToRemove = new HashSet<>();
```

Comment:
a `List` would be more efficient; it is already guaranteed that the `PendingCheckpointId` is unique anyway.